frontend/src/pages/PipelineDetails.test.tsx

```diff
@@ -0,0 +1,328 @@
+    deletePipelineSpy.mockReset();
+    getPipelineSpy.mockImplementation(() => Promise.resolve(testPipeline));
+    getPipelineSpy.mockClear();
+    getTemplateSpy.mockImplementation(() => Promise.resolve('{}'));
+    getTemplateSpy.mockClear();
+    createGraphSpy.mockImplementation(() => new graphlib.Graph());
+    createGraphSpy.mockClear();
+  });
+
+  it('shows empty pipeline details with no graph graph', async () => {
+    TestUtils.makeErrorResponseOnce(createGraphSpy, 'bad graph');
+    const tree = shallow(<PipelineDetails {...generateProps()} />);
+    await getTemplateSpy;
+    await TestUtils.flushPromises();
+    expect(tree).toMatchSnapshot();
+    tree.unmount();
+  });
+
+  it('shows load error banner when failing to get pipeline', async () => {
+    TestUtils.makeErrorResponseOnce(getPipelineSpy, 'woops');
+    const tree = shallow(<PipelineDetails {...generateProps()} />);
+    await getPipelineSpy;
+    await TestUtils.flushPromises();
+    expect(updateBannerSpy).toHaveBeenCalledTimes(2); // Once to clear banner, once to show error
+    expect(updateBannerSpy).toHaveBeenLastCalledWith(expect.objectContaining({
+      additionalInfo: 'woops',
+      message: 'Cannot retrieve pipeline details. Click Details for more information.',
+      mode: 'error',
+    }));
+    tree.unmount();
+  });
+
+  it('shows load error banner when failing to get pipeline template', async () => {
+    TestUtils.makeErrorResponseOnce(getTemplateSpy, 'woops');
+    const tree = shallow(<PipelineDetails {...generateProps()} />);
+    await getPipelineSpy;
+    await TestUtils.flushPromises();
+    expect(updateBannerSpy).toHaveBeenCalledTimes(2); // Once to clear banner, once to show error
+    expect(updateBannerSpy).toHaveBeenLastCalledWith(expect.objectContaining({
+      additionalInfo: 'woops',
+      message: 'Cannot retrieve pipeline template. Click Details for more information.',
+      mode: 'error',
+    }));
+    tree.unmount();
+  });
+
+  it('shows no graph error banner when failing to parse graph', async () => {
+    TestUtils.makeErrorResponseOnce(createGraphSpy, 'bad graph');
+    const tree = shallow(<PipelineDetails {...generateProps()} />);
+    await getTemplateSpy;
+    await TestUtils.flushPromises();
+    expect(updateBannerSpy).toHaveBeenCalledTimes(2); // Once to clear banner, once to show error
+    expect(updateBannerSpy).toHaveBeenLastCalledWith(expect.objectContaining({
+      additionalInfo: 'bad graph',
+      message: 'Error: failed to generate Pipeline graph. Click Details for more information.',
+      mode: 'error',
+    }));
+    tree.unmount();
+  });
+
+  it('shows empty pipeline details with empty graph', async () => {
+    const tree = shallow(<PipelineDetails {...generateProps()} />);
+    await getTemplateSpy;
+    await TestUtils.flushPromises();
+    expect(tree).toMatchSnapshot();
+    tree.unmount();
+  });
+
+  it('collapses summary card when clicking the Hide button', async () => {
```

Comment:
should this also be a snapshot test? is there anything verifying that `summaryShown === true` actually means the card is visible?

ditto below

Reply:
There's a snapshot test above making sure the card is there when state says it's shown.
I'll add another snapshot test to capture when it's not.